### DISCUSSION TITLE
How do I fail a build from a resolver plugin?

### DISCUSSION THREAD

**Discussion by Aaronius** [parcel-bundler/parcel#4992] [General] [answered]
Hello! I'm creating a resolver plugin. When a certain file is required, I would like my plugin to be able to fail the build rather than allow the asset to continue to the next resolver. How would I do this? I can throw a `ThrowableDiagnostic`, which results in a failed build if my resolver is the only resolver in the pipeline, but if there are subsequent resolvers in the pipeline that can resolve the dependency, then my `ThrowableDiagnostic` does not fail the build. How would I accomplish what I'm trying to do? Thanks!

**Comment by DeMoorJasper** [comment, marked as answer]
Currently that's impossible as the way it works is that a thrown error will result in the next resolver getting a chance to resolve the asset as an error usually means that it can't resolve it.

We might add a flag or something similar to it that we can use like the MODULE_NOT_FOUND in the standard node resolver, so we know what kind of error occurred although not entirely sure what the best approach for this is.

**Comment by mischnic** [comment]
> not entirely sure what the best approach for this is.

Making the return type `Promise<Diagnostic | ?ResolveResult>` (for the cases where the default resolver throws) and making throwing errors actually fail the build?

**Reply by DeMoorJasper** [reply]
Yeah, was also thinking about something like this as a possible solution. Maybe we should do this while we're in beta? Would be more expected than a throw doing nothing

**Reply by Aaronius** [reply]
FWIW, throwing a regular error is the first thing I tried. That was before I knew about `ThrowableDiagnostic`, which is the second thing I tried.

**Reply by DeMoorJasper** [reply]
Errors can be converted into diagnostics really easily so it should both work if we end up changing this

**Comment by Aaronius** [comment]
Thanks. Even though it's not the answer I was hoping for, it's definitely helpful. In case it helps to hear our use case, we are creating bundles of untrusted, third-party code (which we call extensions). We want to allow the third-party code to import/require other files that the third-party provided but not any arbitrary file on the server. We also want the third-party code to be able to import/require certain npm packages, but not others. I was hoping to set up my resolvers in my `.parcelrc` like this: `"resolvers": ["parcel-resolver-limit-extension-dependencies","@parcel/resolver-default"],` where `parcel-resolver-limit-extension-dependencies` would be able to fail the build if the extensions depended on something that isn't allowed.

I'll probably either try to call `@parcel/resolver-default` directly from `parcel-resolver-limit-extension-requires` if the thing being required is allowed OR I'll write a babel plugin that detects imports/requires and fails the build if an infraction is found (throwing an error from a Babel plugin does fail the build).

**Reply by DeMoorJasper** [reply]
This seems like a valid use-case, maybe we should change the API to allow this like @mischnic mentioned

**Comment by Aaronius** [comment]
@DeMoorJasper @mischnic Is this being tracked somewhere already or should I create an issue for it?

**Reply by mischnic** [reply]
Here was the issue: https://github.com/parcel-bundler/parcel/issues/5012
PR is already merged
